# PRO
A software package for generating and demultiplexing DNA barcodes in nanopore sequencing.

Information about the paper

## Overview
Hereï¼Œwe develop a software package PRO which can generate and demultiplex barcodes for nanopore sequencing. PRO is based on probability divergence, a "metric" we propose for the dissimilarity between two DNA sequences. For generating barcodes, PRO ensures that any two barcode sequences in a barcode kit are sufficiently distinct from each other. Using PRO, we have greatly expanded the capacity of barcode. For demultiplexing barcodes, PRO achieves higher precision than demultiplexin tools now available for nanopore sequencing in our experiments.

## Installation
    PRO can be used on the linux system.
### Prerequisites
    Anaconda3. For example, users may download and install the following Anaconda3 package:
        wget https://mirrors.tuna.tsinghua.edu.cn/anaconda/archive/Anaconda3-2021.11-Linux-x86_64.sh
        sh Anaconda3-2021.11-Linux-x86 64.sh
### Download the PRO package
        git clone https://github.com/renztr/PRO.git
        cd PRO
### Install required packages
    In order to run PRO,a python interpreter is required (recommended version 3.7.0 and above).In Addition, two essential packages need to be installed: edlib , tqdm. You can run the following command lines:
        pip install edlib==1.3.9
        pip install tqdm    

## Usage
### Generate barcode
    Important parameters for generating barcode
    
    -l	Specify the length of a barcode sequence. 
    -th	Set threshold: the number of edit errors for which a barcode can be correctly demultiplexed. 
    -q	Number of barcodes after pre-processing. Default is 100000.
    -i	A mode of generating barcode: pick barcode set from the user-specified DNA sequences. Please input a txt file containing the specified DNA sequences. Default is None file.
    -f	If barcode needs to keep a certain edit distance from the flanking sequence. Please specify a fastq file contains the flanking sequences used in the library preparation. Default is None file.


    1. Generate barcode from all DNA sequences in k-mer space.
    Python /home/renzt/PRO/pro_FPS_select.py -l 10 -th 2 -o /home/renzt/PRO/examples/testoutput/k_mer_select -q 10000
  
    2. Generate barcode from DNA sequences given by the user-specified DNA sequences.
    python pro_FPS_select.py -l 10 -th 2 -o examples/testoutput/user_select -q 5000 -i examples/testdata/user_barcodes.txt 

### Demultiplex barcode
    Important parameters for demultiplexing barcode

    -i  Directory path containing fasta files of nanopore sequencing results
    -f  A fastq file contains the flanking sequences used in the library preparation.  
    -o  The directory path to save the result files of demultiplexing.
    -t  Specify the number of threads, which affects the speed of demultiplexing. 

## Examples                       
    The examples folder contains the files used to demonstrate the barcode generation. 
    You can run the command line:
        bash /home/renzt/PRO/generate_barcodes.sh
    to generate a barcode set of length 10, which is guaranteed to be demultiplexed correctly if two errors occur in the barcode. Barcodes are generated from the 10,000 DNA sequences that are pre-processed.
    You can run the command line:
        bash /home/renzt/PRO/demultiplexing.sh
    to perform demultiplexing. The sequencing data in examples folder file is generated by the simulator deepsimulator 1.5.

    Explanation of the Content in the Output Folder:
    When selecting the barcode set, we first pre-process the input DNA sequences and output two files: pre_processing_barcodes.txt and pre_processing_barcodes_number.txt, which are the pre-processed DNA sequences and the numbers corresponding to these sequences. The final select_barcode.txt is then sampled by farthest point for barcode selection.The folder dex_result contains the results of the demultiplexing.